### PR TITLE
tortoise evict under pbase

### PR DESCRIPTION
**Pull request by almogdepaz**

### CODE REVIEW

File: tortoise/ninja_tortoise.go [hunk 1]
```diff
@@ -218,15 +231,15 @@ func (ni *ninjaTortoise) updateCorrectionVectors(p votingPattern, bottomOfWindow
 				panic(fmt.Sprintf("error block not found ID %d", bid))
 			}
 
-			if _, found := ni.tExplicit[b.Id][x.Layer()]; found { //if Texplicit[b][x]!=0 check correctness of x.layer and found
+			if _, found := ni.tExplicit[b.ID()][x.Layer()]; found { //if Texplicit[b][x]!=0 check correctness of x.layer and found
```
Comment (barakshani) [inline tortoise/ninja_tortoise.go]: in the code comment, x should be layer(x)

File: mesh/meshdb.go [hunk 2]
```diff
@@ -171,27 +171,24 @@ func (mc *MeshDB) ForBlockInView(view map[types.BlockID]struct{}, layer types.La
 		block, err := mc.GetMiniBlock(a)
 		if err != nil {
 			//todo: add err returned from handler, if error is returned return it, if not - keep going
```
Comment (zalmen) [inline mesh/meshdb.go]: remove todo

File: tortoise/ninja_tortoise.go [hunk 3]
```diff
@@ -106,10 +108,45 @@ func NewNinjaTortoise(layerSize int, blocks BlockCache, log log.Log) *ninjaTorto
...
 }
 
-func (ni *ninjaTortoise) processBlock(b *types.Block) {
+func (ni *ninjaTortoise) evictOutOfPbase(old types.LayerID) {
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	for i := old; i < ni.pBase.Layer(); i++ {
+		go func() {
+			defer wg.Done()
+			for _, i := range ni.patterns[i] {
```
Comment (zalmen) [inline tortoise/ninja_tortoise.go]: you are using `i` twice, replace one of them

File: tortoise/ninja_tortoise.go [hunk 4]
```diff
@@ -106,10 +108,45 @@ func NewNinjaTortoise(layerSize int, blocks BlockCache, log log.Log) *ninjaTorto
 	}
 }
 
-func (ni *ninjaTortoise) processBlock(b *types.Block) {
+func (ni *ninjaTortoise) evictOutOfPbase(old types.LayerID) {
+	wg := sync.WaitGroup{}
+	wg.Add(1)
```
Comment (zalmen) [inline tortoise/ninja_tortoise.go]: should be inside the for loop

File: tortoise/ninja_tortoise.go [hunk 5]
```diff
@@ -106,10 +108,45 @@ func NewNinjaTortoise(layerSize int, blocks BlockCache, log log.Log) *ninjaTorto
...
+	}
+}
 
-	ni.Debug("process block: %d layer: %d  ", b.Id, b.Layer())
+func (ni *ninjaTortoise) processBlock(b *types.Block) {
 
+	ni.Debug("process block: %d layer: %d  ", b.ID(), b.Layer())
 	if b.Layer() == Genesis {
 		return
 	}
```
Comment (zalmen) [inline tortoise/ninja_tortoise.go:152]: Comment regarding old code below - in the `ni.Error` few lines below, you should also print the error

**Review by zalmen — CHANGES_REQUESTED** [review]

**Review by zalmen — APPROVED** [review]

**Pull request merged**